### PR TITLE
ref(hybrid-cloud): Add user services. Start tagging some model tests as stable

### PR DESCRIPTION
**Pull request by mikejihbe**
Start tagging some model tests as stable. Implement first "in the wild" service interface and start replacing some usage.

### CODE REVIEW

File: src/sentry/services/hybrid_cloud/users.py
Comment (corps) [inline src/sentry/services/hybrid_cloud/users.py:49]: The approach I took with organization, and membership, was to create new data classes and not use the existing ORM objects.  My thinking was that doing so would help with
1. Clarifying that these objects generally should not have further ORM object calls, such as on relations, or accidental saves.
2. Explicitly stating which fields were actually needed in the cross silo data case, so that when an api or alternative implementation was actually made, there was stronger clarity on which fields would be needed.

Do you imagine that a future pass at this breaks out an `ApiUser` data class object instead of the ORM?  It has implications across the codebase but I feel that it brings a lot of clarity with the other such objects.

Comment (mikejihbe) [reply src/sentry/services/hybrid_cloud/users.py:49]: That makes total sense. I hadn't gotten that far, to be honest. My plan was to identify all the interface points to extract these interfaces, and then come back later to revisit the actual transport layer. Depending on how we implement that, it seemed sort of likely that there might be automated data classes generated from that work.

Comment (corps) [reply src/sentry/services/hybrid_cloud/users.py:49]: Yup, that makes fair sense.   As long as we can come back around and lock that stuff down, extracting atleast the interface points is important.

File: src/sentry/models/actor.py
Comment (markstory) [inline src/sentry/models/actor.py:26]: I think we use `cls` more than `klass`.

File: src/sentry/services/hybrid_cloud/user.py
Comment (markstory) [inline src/sentry/services/hybrid_cloud/user.py:32]: Could be `get_many` 🚲 🏠

File: src/sentry/services/hybrid_cloud/user.py
Comment (markstory) [inline src/sentry/services/hybrid_cloud/user.py:68]: If we're going to have both the db and remote implementations return the same thing we shouldn't be returning querysets as it is unlikely that we would build a queryset implementation for the RPC/cache backend. Should we cast to `list()` here as a quick fix?

Comment (mikejihbe) [reply src/sentry/services/hybrid_cloud/user.py]: I did this for now. After some discussion with @corps, we actually think it may make sense to return ORM models in some cases (particularly where we expect a read-through cache), but to make the services return totally distinct models per silo with corresponding distinct ORM relationships. This gives total clarity when dealing with each different model

Comment (corps) [reply src/sentry/services/hybrid_cloud/user.py]: I mean, hmm.
One thing, that would be useful @mikejihbe is if you did, still, extract a dataclass *even if we implement that dataclass to look a lot like an ORM* (for instance, adding helper properties that access the relation, etc).

I think it's possible to do this without colouring the entire world, but it starts to enforce certain guarantees of access.  When we're using the `user_service` in other places in the code, we should be informed when the impact of pulling objects from the service might change which relationships and other access we have based on the silo.

Even if we're going to use 2 different ORM classes in the future, I do think there is value whilst laying these APIs out to actually atleast split the class out from the existing ORM one.  Would it be possible you think to return something for now that you either 1.  forcibly upgrade into real User orm objects or 2.  is duct typed minimally to represent the needs of your callsites?

Comment (corps) [reply src/sentry/services/hybrid_cloud/user.py]: I'm strugging between realities of how useful it is to get the interface extracted before we know what our final typing is going to be (or what the final RPC / caching layer looks like) with also the sense that dividing the types and seeing the deeper impact has been useful broadly for me, too.  Did you have further thoughts @markstory ?

I don't want this PR or any PR to become huge, but I'm pondering if there is a compromise.

File: src/sentry/models/actor.py
Comment (corps) [inline src/sentry/models/actor.py:38]: For instance, this could just be a `User.objects.get(user.id)`.  That would allow our upstream type to be anything, and we can still later decide to down or up grade this code path based on how we chose to handle the caching layer.

File: src/sentry/services/hybrid_cloud/user.py
Comment (corps) [reply src/sentry/services/hybrid_cloud/user.py]: I think my final answer is this:
In the end, I think we want to have a smooth ORM like interface for typical developers.  I think that interface, however, does not exist purely on the service layer object.  I think in the end another layer helps translate to an ORM object that is local to the region.  Ideally that layer is invoked closer to controllers and *not* inside common helpers.

The "big problem" I want to avoid that can be insidious is helper functions that are written using any monolith ORM object that may unfortunately only have test coverage in one test silo by the end, but still be invoked and fail in the other silo.  

I want developers to see and understand that we are "colouring" the app across silo bounds and that yes, there is atleast some substantial enough difference in access now that they should be thinking about.  I want that to be obvious and I want it to be testable without it being silent.

Comment (corps) [reply src/sentry/services/hybrid_cloud/user.py]: What that means for this particular PR is that I would prefer the user service return dumb objects, and we explicitly upgrade them into ORM objects.  That way when we expand on the usage of the user service we are forced to see the impact and start considering what the next step is ensuring that our system is still performant and easy to use.

Comment (corps) [reply src/sentry/services/hybrid_cloud/user.py]: Or maybe not.  The User model may benefit in the end from consistent interface.  Arrrgh.  I don't know.  🤷   I literally jsut keep going back and forth.

Comment (markstory) [reply src/sentry/services/hybrid_cloud/user.py]: I think focusing on the resulting workflow is important as we'll be using these interfaces and decisions for a while. I'm of two minds still as well.

Returning ORM entities from the services is convenient for us in the present and in future scenarios when data is 'local' to the silo operations are being done it. Folks will generally be comfortable with using ORM entities and doing things like fetching relations. Using ORM entities are problematic in the multi-region world because of how powerful and flexible they are. We simply won't be able to do `user.save()` in most application code in the future. Instead we'll have to go through the service abstractions to have consistent behavior.

Using Data transfer objects or dumb entities is attractive as we can make both the local and remote implementations of the service the same, but we need to re-implement all the logic that User ORM objects have.

I'm partial to using ORM entities. My thinking is that:

1. We're already patching object managers to raise errors on incorrect silo use.
2. Using ORM objects will let us disrupt other teams less.
3. We don't need to come up with a solid abstraction for entities as we're using the one that django provides.

One looming concern with ORM entities is how we handle relations. The ORM allows incremental lazy loading through related managers and we simply don't have that luxury with remote service implementations.

File: src/sentry/services/hybrid_cloud/user.py
Comment (corps) [inline src/sentry/services/hybrid_cloud/user.py:25]: Interestingly, I added this into the base service.  Checkout `_serialize_organization`.  But that was because I needed some additional custom logic.  I have no opinion beyond that.

File: src/sentry/services/hybrid_cloud/user.py
Comment (corps) [inline src/sentry/services/hybrid_cloud/user.py:29]: This totally works, but there's an even more elegant way:

https://docs.python.org/3/library/dataclasses.html#dataclasses.fields

You can even add custom metadata to 'hint' at how unpacking should be done.

File: src/sentry/services/hybrid_cloud/user.py
Comment (corps) [reply src/sentry/services/hybrid_cloud/user.py]: I'm putting together a doc on this subject in greater detail where we can locate further details on this subject.

I definitely was in favor of ORM entities in the case of Region resources initially, but my opinion on that has been evolving.  I encouraged mike to try using data classes here but it doesn't have to be a final decision.

Can I give one doc where I articulate, much better than I have, a definitive reason for the service class not to expose ORM objects?  Then we can discuss and come to the conclusion together.

File: src/sentry/services/hybrid_cloud/user.py
Comment (corps) [inline src/sentry/services/hybrid_cloud/user.py:44]: This should actually be renamed and clarified.  Based on its usage, it looks like the point of this is to fetch *case insensitive email matches* specifically.  There should only be one user per email, but there can be multiple users per case insensitive emails.

Comment (mikejihbe) [reply src/sentry/services/hybrid_cloud/user.py:44]: There wasn't actually any usage of case sensitive fetching of emails, so I removed the bifurcation to reflect actual usage

Comment (corps) [reply src/sentry/services/hybrid_cloud/user.py:44]: Oh snap, well cool.  👍   More clarity is good.

File: src/sentry/api/serializers/models/actor.py
Comment (corps) [inline src/sentry/api/serializers/models/actor.py:10]: Would it make sense for `get_display_name` to simply become a `staticmethod` that is provided arguments it needs?  That prevent the need for this helper to be duplicated.

File: src/sentry/services/hybrid_cloud/user.py
Comment (corps) [inline src/sentry/services/hybrid_cloud/user.py:105]: Why is this conditional possible?  The typing suggests it should not be.

File: src/sentry/services/hybrid_cloud/user.py
Comment (corps) [inline src/sentry/services/hybrid_cloud/user.py:38]: There will come a time I think when this logic has to go and be replaced with explicit attribute packing.  The problem is that not all attributes will carry the same typing by the end, a good example being relationships, but some others as well.

File: src/sentry/services/hybrid_cloud/user.py
Comment (mikejihbe) [reply src/sentry/services/hybrid_cloud/user.py:105]: Good catch. The typing was wrong.

File: src/sentry/services/hybrid_cloud/user.py
Comment (mikejihbe) [reply src/sentry/services/hybrid_cloud/user.py:38]: I'm not sure I agree.
Relationship attributes are named without the `_id` suffix. E.g. `.actor` is the relationship, `.actor_id` is the data value. So we just copy the data values, [e.g.](https://github.com/getsentry/sentry/pull/40614/files/0441a3c3579dbf742bd6f03fdee699db77bdea84#diff-5ee4410f00e99e25c0534d979882b2211d87337651d4083076f9af13340179caR24)

File: src/sentry/models/groupsubscription.py
Comment (corps) [inline src/sentry/models/groupsubscription.py:123]: Hold on, you can actually avoid the service boundary altogether and avoid needing the `get_from_group` call.  That's because you only need the `user_id`, you don't need any other attribute of the user, and all the actually relevant models are in the region silo.

What about splitting `User.objects.get_from_group` up into two steps: one step , you find the org membership (which is local to the region) and return the relation (user_id, just the number).  The second step, fetches the users based on those user ids.

This greatly simplifies all this logic and prevents the unnecessary silo exchange (which become expensive in the end).

Comment (corps) [reply src/sentry/models/groupsubscription.py:123]: Ah, that involves a lot of other code changes as well.
This seems brittle, hmm hmm hmm.
A LOT of time we use the full ORM object when all we mean is an id, and it's very very frustrating.

File: src/sentry/services/hybrid_cloud/user.py
Comment (corps) [reply src/sentry/services/hybrid_cloud/user.py:38]: You mentioned you wanted to avoid this becoming "too rpc-y".  As a result, I've been refactoring in such a way that many domain relations come "packed" into an object, which is going to be much more performant and correct in the end.  In that case, something like this cannot possibly work:

https://github.com/getsentry/sentry/pull/40772/files#diff-4c7c9fddec7c4378c1d12c024acfd6901b6c74750350cda751ef79970bbd7d70R90

File: src/sentry/models/groupsubscription.py
Comment (mikejihbe) [reply src/sentry/models/groupsubscription.py:123]: Yeah, the "recipient" type switch matters in a _lot_ of places. Splitting up the notification system based on type would help a lot

File: src/sentry/models/organization.py
Comment (corps) [inline src/sentry/models/organization.py:254]: Seems like the return type needs to change, `Sequence[ApiUser]`.

In the future, this could just go directly on the `organization_service` to be honest.  I'll take a pass at it after your work lands. by pulling it into my PR.

File: src/sentry/models/team.py
Comment (corps) [inline src/sentry/models/team.py:167]: Is there a reason we don't use `type(self).__name__`?  Just curious it's not really a big issue.

File: src/sentry/notifications/notifications/base.py
Comment (corps) [inline src/sentry/notifications/notifications/base.py:175]: AH, now I see.  Ok, hmmmmmmmmmm man that sucks but ok 🤷

File: src/sentry/services/hybrid_cloud/user.py
Comment (corps) [inline src/sentry/services/hybrid_cloud/user.py:33]: Does it make sense for this to just be a static property on APIUser?  And let the control silo pack it in with its own logic?

Thing is, I can see API drift causing issues with this.  We deploy control silo and it shows one name for users, but it's not in agreement with regions until they are all fully deployed.

If Control silo is source of truth for these sorts of "computed properties", I'd argue that at some point they should be static properties and determined as their source only.  I'm ok with this for now, but I'd atleast put a huge "KILL ME" there.

File: src/sentry/services/hybrid_cloud/user.py
Comment (mikejihbe) [reply src/sentry/services/hybrid_cloud/user.py:38]: Ah yes, this is definitely true.

File: src/sentry/api/serializers/models/actor.py
Comment (markstory) [reply src/sentry/api/serializers/models/actor.py:10]: Could be a good standalone function for the user service to expose as well.

File: src/sentry/services/hybrid_cloud/user.py
Comment (markstory) [reply src/sentry/services/hybrid_cloud/user.py:33]: Having these computed properties come from Control Silo and be pat of the User 'dataclass' seems like a good direction to go.

**Comment by corps** [comment]
Getsentry side: https://github.com/getsentry/getsentry/pull/8754

File: src/sentry/api/fields/actor.py
Comment (corps) [inline src/sentry/api/fields/actor.py:31]: `ActorTuple.type` is still User in either case, making this comparison safe.

File: src/sentry/api/fields/actor.py
Comment (corps) [inline src/sentry/api/fields/actor.py:24]: Wherever I thought it was helpful to really deeply annotate what's going on, I added more annotations.

File: src/sentry/api/invite_helper.py
Comment (corps) [inline src/sentry/api/invite_helper.py:229]: Hmm, @mikejihbe Is this change going to truly be necessary?  I think we should be able to revert this, actually.

File: src/sentry/api/serializers/models/actor.py
Comment (corps) [reply src/sentry/api/serializers/models/actor.py:9]: I can't, quite yet just reference `ApiUser` as there are some ActorSerializer callsites that do reference raw ORMs still.  This compromise remains until we can close that hole.

File: src/sentry/api/serializers/models/group.py
Comment (corps) [inline src/sentry/api/serializers/models/group.py:185]: This is semi faithful move of `ActorTuple.resolve_dict` which was a massive misnomer.  That method was only referenced here, nowhere else, and what it *actually* does is try to return a dictionary of group_ids -> serializable entities.  This new method captures most of that logic but uses the new user_service.serialize_users rather than expect raw Users which implicitly get serialized.  Keeping it in this serializer class also makes clear its specific, non generic intent.

File: src/sentry/api/serializers/models/group.py
Comment (corps) [inline src/sentry/api/serializers/models/group.py:229]: group_id -> actor tuple, not actors.  The typing on this was wrong originally (see `resolve_dict`)

File: src/sentry/api/serializers/rest_framework/mentions.py
Comment (corps) [inline src/sentry/api/serializers/rest_framework/mentions.py:42]: This could probably just be `APIUser` actually.  We only need `BaseUser` for codepaths that risk taking both ORMs and our dataclasses, but this function is utility and only comes from one callsite.

File: src/sentry/api/validators/project_codeowners.py
Comment (corps) [inline src/sentry/api/validators/project_codeowners.py:74]: It actually could remain `get_display_name`, perhaps I should revert.

File: src/sentry/auth/helper.py
Comment (corps) [inline src/sentry/auth/helper.py:428]: Hmm, feeling this also is probably not the best spot for BaseUser.  I'll revert.

File: src/sentry/models/organization.py
Comment (corps) [reply src/sentry/models/organization.py]: Replaced this with `ApiUser` and made downstream changes.

File: src/sentry/models/actor.py
Comment (corps) [inline src/sentry/models/actor.py:95]: Trashing this

File: src/sentry/models/group.py
Comment (corps) [inline src/sentry/models/group.py:463]: If I'm honest, I'm not sure what this change, is.  @mikejihbe ?

File: src/sentry/api/invite_helper.py
Comment (mikejihbe) [reply src/sentry/api/invite_helper.py:229]: it might not be. It seemed safer to over-apply the BaseUser check than to under apply it, so that's what I did.

File: src/sentry/models/group.py
Comment (mikejihbe) [reply src/sentry/models/group.py:463]: This removed a deprecation error I ran into. Unrelated to this changeset.

File: src/sentry/notifications/notifications/base.py
Comment (mikejihbe) [inline src/sentry/notifications/notifications/base.py:89]: @corps are these still valid types now that I put the User up-class in here? Does that matter for now?

Comment (corps) [reply src/sentry/notifications/notifications/base.py:89]: This should still be fine to send out.  I validated that the code paths worked for APIUser and User simult.  There could be one or two exceptions where I made a very minor change to support APIUser, so just check function body.

If function body is unchanged, it's safe to keep and it means I verified it by hand working the same for both types.

File: src/sentry/api/serializers/base.py
Comment (corps) [inline src/sentry/api/serializers/base.py:59]: I think this isn't necessary, actually.  The reasoning is that in the case that there is not an explicit serializer, the dict will still be pass through.  I have an alternative fix to the issue you were seeing, hold on.

File: src/sentry/api/serializers/models/group.py
Comment (corps) [inline src/sentry/api/serializers/models/group.py:198]: So I now realize why I was confused before.  I was focused on the fact this is shoved in `assigned_to` but was missing the fact that this ended up in the `serialize` call with an explicit Serializer.  So I made two changes trying to fix the same issue, but this change isn't necessary actually.  Here's my suggestion:

Try removing this conditional.  `resolved_actors[t] = {actor.id: actor for actor in serializable}` is sufficient for both cases (no need for the serialize_users call, I was wrong).  Then try removing your conditional to the serializer.  I think that should work due to the `ActorSerializer` changes.

File: src/sentry/digests/utils.py
Comment (mikejihbe) [inline src/sentry/digests/utils.py:73]: @corps this is the only weird change I had to make to get all these changes working. Some tests are still getting users into  `participants_by_provider_by_event`

**Review by corps — APPROVED** [review]

**Pull request merged**
